### PR TITLE
Homogenous tests + Heterogenous tests + minor bug fixes + output changes

### PR DESCRIPTION
**Pull request by nimeshnischal**
Sorry for the huge diff!

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -10,4 +10,5 @@ RUN tar xzvf apache-maven-3.8.6-bin.tar.gz -C/opt/
 ENV PATH="${PATH}:/opt/apache-maven-3.8.6/bin"
 
 RUN apt-get -y install sudo
-VOLUME /submissions
+
+VOLUME ["/root/.m2","/submissions"]
```
Comment (breed) [inline Dockerfile:14]: we don't want to hardcode source paths since we want to set those at runtime in the .config file

File: codeval.py [hunk 2]
```diff
@@ -1,42 +1,21 @@
...
 import click
 import datetime
 import os, shutil, sys
-import requests
-import zipfile
 import subprocess
 import tempfile
-import time
 import traceback
+from commons import *
```
Comment (breed) [inline codeval.py:9]: can you expand this import? it's good to be clear on exactly what we are importing from where.

File: codeval.py [hunk 3]
```diff
@@ -54,7 +33,7 @@ def __init__(self):
             user = self.canvas.get_current_user()
             info(f"connected to canvas as {user.name} ({user.id})")
         except:
-            error(f"there was a problem accessing canvas.")
+            error(f"there was a problem accessing canvas.", True)
```
Comment (breed) [inline codeval.py:36]: why is there a Turue?

File: commons.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+def _now():
+    return time.strftime("%Y-%m-%d %H:%M:%S", time.gmtime())
+
+
+def debug(message):
+    if get_config().show_debug:
+        click.echo(click.style(f"{_now()} D {message}", fg='magenta'))
+
+
+def error(message, raise_exception=False):
```
Comment (breed) [inline commons.py:38]: is the raise_exception to make the program exit? i think it's better to add a fatal level and make that exit

**Review by breed — CHANGES_REQUESTED** [review]
just a couple of minor changes. also can you add a README to describe how to run it?

File: commons.py (hunk 4 above)
Comment (nimeshnischal) [reply commons.py]: Separated error(message) and errorWithException(message)

File: codeval.py (hunk 3 above)
Comment (nimeshnischal) [reply codeval.py]: Removed True and replaced with errorWithException()

File: codeval.py (hunk 2 above)
Comment (nimeshnischal) [reply codeval.py]: Expanded all import *

**Review by breed — APPROVED** [review]
thank you nimesh

**Pull request merged**